lib/remote_node_indexer.go

```diff
@@ -0,0 +1,237 @@
+	RemoteNodeIndexId_NonValidator_Attempted RemoteNodeIndexId = "NONVALIDATOR_ATTEMPTED"
+)
+
+// RemoteNodeIndexer is a structure that holds information about all remote nodes and their indices.
+type RemoteNodeIndexer struct {
+	// AllRemoteNodes is a map storing all remote nodes by their IDs.
+	AllRemoteNodes *RemoteNodeIndex[RemoteNodeId]
+
+	// Indices for various types of remote nodes.
+	ValidatorIndex             *RemoteNodeIndex[bls.PublicKey]
+	ValidatorAttemptedIndex    *RemoteNodeIndex[bls.PublicKey]
+	NonValidatorOutboundIndex  *RemoteNodeIndex[RemoteNodeId]
+	NonValidatorInboundIndex   *RemoteNodeIndex[RemoteNodeId]
+	NonValidatorAttemptedIndex *RemoteNodeIndex[RemoteNodeId]
+
+	// RemoteNodeToIndexIdList maps remote nodes to their corresponding index IDs.
+	RemoteNodeToIndexIdList map[*RemoteNode][]RemoteNodeIndexId
+
+	RemoteNodeToIndexRemoveFuncList map[*RemoteNode][]RemoteNodeIndexRemoveFunc
+}
+
+// NewRemoteNodeIndexer initializes and returns a new instance of RemoteNodeIndexer.
+func NewRemoteNodeIndexer() *RemoteNodeIndexer {
+	rni := &RemoteNodeIndexer{
+		RemoteNodeToIndexIdList:         make(map[*RemoteNode][]RemoteNodeIndexId),
+		RemoteNodeToIndexRemoveFuncList: make(map[*RemoteNode][]RemoteNodeIndexRemoveFunc),
+	}
+
+	// Initializing various indices with their respective types and update callback.
+	rni.AllRemoteNodes = NewRemoteNodeIndex[RemoteNodeId](RemoteNodeIndexId_Default, rni.updateCallback)
+	rni.ValidatorIndex = NewRemoteNodeIndex[bls.PublicKey](RemoteNodeIndexId_Validator, rni.updateCallback)
+	rni.ValidatorAttemptedIndex = NewRemoteNodeIndex[bls.PublicKey](RemoteNodeIndexId_ValidatorAttempted, rni.updateCallback)
+	rni.NonValidatorOutboundIndex = NewRemoteNodeIndex[RemoteNodeId](RemoteNodeIndexId_NonValidator_Outbound, rni.updateCallback)
+	rni.NonValidatorInboundIndex = NewRemoteNodeIndex[RemoteNodeId](RemoteNodeIndexId_NonValidator_Inbound, rni.updateCallback)
+	rni.NonValidatorAttemptedIndex = NewRemoteNodeIndex[RemoteNodeId](RemoteNodeIndexId_NonValidator_Attempted, rni.updateCallback)
+	return rni
+}
+
+// Getter methods for accessing the different indices.
+func (rni *RemoteNodeIndexer) GetAllRemoteNodes() RemoteNodeIndexInterface[RemoteNodeId] {
+	return rni.AllRemoteNodes
+}
+
+func (rni *RemoteNodeIndexer) GetValidatorIndex() RemoteNodeIndexInterface[bls.PublicKey] {
+	return rni.ValidatorIndex
+}
+
+func (rni *RemoteNodeIndexer) GetValidatorAttemptedIndex() RemoteNodeIndexInterface[bls.PublicKey] {
+	return rni.ValidatorAttemptedIndex
+}
+
+func (rni *RemoteNodeIndexer) GetNonValidatorOutboundIndex() RemoteNodeIndexInterface[RemoteNodeId] {
+	return rni.NonValidatorOutboundIndex
+}
+
+func (rni *RemoteNodeIndexer) GetNonValidatorInboundIndex() RemoteNodeIndexInterface[RemoteNodeId] {
+	return rni.NonValidatorInboundIndex
+}
+
+func (rni *RemoteNodeIndexer) GetNonValidatorAttemptedIndex() RemoteNodeIndexInterface[RemoteNodeId] {
+	return rni.NonValidatorAttemptedIndex
+}
```

Comment:
This comment is mostly copy-pasted from https://github.com/deso-protocol/core/pull/858 and adapted based on our offline chats. Nothing significantly new here.

This comment hinges on the suggestion to merge (NonValidatorOutboundIndex and NonValidatorAttemptedIndex) into a single map, and to merge (ValidatorIndex, ValidatorAttemptedIndex) into a single map. This way, you would have a total of three maps:
- ValidatorIndex map[bls.PublicKey]*RemoteNode
   - This would merge the existing ValidatorAttemptedIndex and ValidatorIndex
- NonValidatorOutboundIndex map[RemoteNodeId]*RemoteNode
  - This would merge the existing NonValidatorAttemptedIndex and NonValidatorOutboundIndex
- NonValidatorInboundIndex map[RemoteNodeId]*RemoteNode
If the suggestion to merge the maps works, then you can simplify the indexer by exposing separate Add/Remove/Get functions for the three node types (Validator, NonValidatorOutbound, NonValidatorInbound). There would be no need for the separate AllRemoteNodes map.

It'll eliminate the need for the callback function, the extra complexity in RemoveRemoteNode to iterate through all of the indices, and the RemoteNodeIndex templated struct itself. The code overall becomes easier to read and simpler to understand, even if does not look as fancy. It's < 3 hours of work to apply this code review feedback.

I know you personally may not like this suggested solution given it's less generic and can be less computationally optimal. That said, please give this change an honest consideration in the spirit of optimizing for understandability and simplicity. What most important is that the code allows for O(1) lookup of validators and non-validators from the critical paths in consensus. The number of remote nodes will be small in the medium term ( < 1000 ), so asymptotic time complexity isn't a huge worry for fetching all remote nodes of a certain type. 

```suggestion
func (rni *RemoteNodeIndexer) AddValidator(node *RemoteNode) {
	rni.ValidatorIndex[node.PublicKey] = node
}
func (rni *RemoteNodeIndexer) RemoveValidator(node *RemoteNode) {
	delete(rni.ValidatorIndex, node.HandshakeMetadata.validatorPublicKey)
}
func (rni *RemoteNodeIndexer) GetValidator(pk *bls.PublicKey) *RemoteNode {
	node, _ := rni.ValidatorIndex[node.HandshakeMetadata.validatorPublicKey]
	return node
}
func (rni *RemoteNodeIndexer) GetAllValidators() []*RemoteNode {
	return collections.MapValues(rni.ValidatorIndex)
}
func (rni *RemoteNodeIndexer) AddNonValidator(node *RemoteNode) {
	// Check if the node is outbound vs. inbound and add it to the appropriate map
}
func (rni *RemoteNodeIndexer) RemoveNonValidator(node *RemoteNode) {
	// Remove the node from the outbound or inbound non-validator index: whichever one it's stored in
}
func (rni *RemoteNodeIndexer) GetNonValidator(id RemoteNodeID) *RemoteNode {
	// Fetch the node from the outbound or inbound non-validator index: whichever one it's stored in
}
func (rni *RemoteNodeIndexer) GetAllNonValidators() []*RemoteNode {
	// Return a slice of all remote nodes from the outbound and inbound non-validator indices
}
func (rni *RemoteNodeIndexer) GetAllRemoteNodes() []*RemoteNode {
	// Return a slice of all nodes from the three indices
}
```